x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pytorch/process/PyTorchResultProcessor.java

```diff
@@ -113,20 +140,66 @@ private void processInferenceResult(PyTorchInferenceResult inferenceResult) {
     }
 
     public synchronized ResultStats getResultStats() {
+        long currentMs = currentTimeMsSupplier.getAsLong();
+        long currentPeriodStartTimeMs = startTime + Intervals.alignToFloor(currentMs - startTime, REPORTING_PERIOD_MS);
+
+        // Do we have results from the previous period?
+        RecentStats rs = null;
+        if (lastResultTimeMs >= currentPeriodStartTimeMs) {
+            // if there is a result for the last period then set it.
+            // lastPeriodStats will be null when more than one period
+            // has passed without a result.
+            rs = lastPeriodStats;
+        } else if (lastResultTimeMs >= currentPeriodStartTimeMs - REPORTING_PERIOD_MS) {
+            // there was a result in the last period but not one
+            // in this period to close off the last period stats.
+            // The stats are valid return them here
+            rs = new RecentStats(lastPeriodSummaryStats.getCount(), lastPeriodSummaryStats.getAverage());
+        }
+
+        if (rs == null) {
+            // no results processed in the previous period
+            rs = new RecentStats(0L, 0.0);
+        }
+
         return new ResultStats(
             new LongSummaryStatistics(timingStats.getCount(), timingStats.getMin(), timingStats.getMax(), timingStats.getSum()),
             errorCount,
             pendingResults.size(),
-            lastUsed
+            Instant.ofEpochMilli(lastResultTimeMs),
+            this.peakThroughput,
+            rs
         );
     }
 
     private synchronized void processResult(PyTorchInferenceResult result) {
-        if (result.isError() == false) {
-            timingStats.accept(result.getTimeMs());
-            lastUsed = Instant.now();
-        } else {
+        if (result.isError()) {
             errorCount++;
+            return;
+        }
+
+        timingStats.accept(result.getTimeMs());
+
+        lastResultTimeMs = currentTimeMsSupplier.getAsLong();
```

Comment:
Seems to me the `PyTorchInferenceResult` should be the thing that keeps track of the result time. 

But that is a nit, and such a small optimization (inference time still dominates, so looking up system time is still very fast comparatively).